### ISSUE TITLE
Return a view from tensor(requires_grad=False) in autograd function may cause incorrect requires_grad attribute.

### ISSUE THREAD

**Issue by Ghost-LZW** [pytorch/pytorch#166131]
### 🐛 Describe the bug

```python
import torch

a = torch.rand(4, requires_grad=True)

b = a.view(4)
assert b.requires_grad and b._is_view()

with torch.no_grad():
    c = b.view(4)

assert c.requires_grad

class View(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        return x.data.view(x.shape)

    @staticmethod
    def backward(ctx, grad):
        return grad

d = View.apply(a)
assert d.requires_grad and d._is_view()

with torch.no_grad():
    e = d.view(4)

assert e.requires_grad # AssertionError
```

The variable `d` in the `forward` method of other `autograd.Function` implementations may behave differently compared to `b`. But I don’t have any method to distinguish between them.

### Versions

Versions of relevant libraries:
[pip3] numpy==2.3.4
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] pytorch-triton==3.5.0+git7416ffcb
[pip3] torch==2.10.0.dev20251023+cu126
[pip3] torchvision==0.25.0.dev20251023+cu126

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @xmfan

**Comment by Ghost-LZW** [comment]
@albanD Using `.detach()` can also reproduce this issue, while `x.view(x.shape)` seems to resolve it. I think the underlying mechanism might be that if an autograd function returns a view tensor, subsequent operations performed in a `no_grad` context will inherit the `requires_grad` attribute from the original tensor. This can lead to unexpected behavior during gradient propagation.

**Comment by albanD** [comment]
Why would you expect `e` to require gradients here given that it was created in no_grad mode?
I agree that it is inconsistent with the non-custom Function case though indeed.

The reason is that "detach()" is a non-differentiable view. And so the output of the custom Function is not considered a view of its input. Leading to slightly different handling downstream.

**Comment by Ghost-LZW** [comment]
> Why would you expect `e` to require gradients here given that it was created in no_grad mode? 

The examples are extracted from real use case, where they actually occur within autograd.Function rather than under torch.no_grad. In this implementation, the decision to retain activations relies on the requires_grad propagation introduced by the view operation in no_grad mode.
